File: binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        parent_nodes = find_parents(root, p)
        curr = root
        res = curr
        while True:
            if curr.val == q.val:
                if q.val in parent_nodes:
                    return q
                else:
                    return res
            elif curr.val > q.val:
                curr = curr.left
            else:
                curr = curr.right
            if curr.val in parent_nodes:
                res = curr
def find_parents(root, node):
    res = set([root.val])
    while True:
        if root.val == node.val:
            res.add(node.val)
            return res
        elif root.val < node.val:
            root = root.right
        else:
            root = root.left
        res.add(root.val)
    return res

File: binary-tree/lowest-common-ancestor-of-a-binary-search-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

def dfs(node, target_node, res)
    return if node.nil?
    if node.val == target_node.val
        res.append(node)
        return res
    end
    if dfs(node.left, target_node, res)
        res.append(node)
        return res
    end
    if dfs(node.right, target_node, res)
        res.append(node)
        return res
    end
end

def get_path(root, target_node)
    res = []
    dfs(root, target_node, res)
    res
end

# @param {TreeNode} root
# @param {TreeNode} p
# @param {TreeNode} q
# @return {TreeNode}
def lowest_common_ancestor(root, p, q)
    p_set = Set.new(get_path(root, p))
    q_path = get_path(root, q)
    q_path.each do |node|
        if p_set.include? node
            return node
        end
    end
end